src/modules/formBuilder/components/FormBuilderPage.tsx

```diff
@@ -1,86 +1,116 @@
+    onCompleted: (data) => {
+      if (
+        data.updateFormDefinition?.errors &&
+        data.updateFormDefinition.errors.length > 0
+      ) {
+        setErrorState(partitionValidations(data.updateFormDefinition.errors));
+      } else {
+        setWorkingDefinition(
+          data.updateFormDefinition?.formDefinition?.definition
+        );
+        setErrorState(emptyErrorState);
+        closeItemEditor();
+      }
+    },
+  });
+
+  const formDefinition = useMemo(() => {
+    if (data && data.updateFormDefinition?.formDefinition)
+      return data.updateFormDefinition.formDefinition;
+    return initialFormDefinition;
+  }, [data, initialFormDefinition]);
+
+  if (fetchError) throw fetchError;
+  if (saveError) throw saveError;
+
+  if (!formDefinition) {
+    if (fetchLoading) return <Loading />;
+    return <NotFound />;
+  }
 
-  if (error) throw error;
-  if (!formDefinition) return <Loading />;
+  // TODO(#6094, #6083) Disable interaction with the form structure while the save mutation is loading
 
   return (
-    <Box sx={{ display: 'flex' }}>
-      <FormBuilderPalette />
-      {selectedItem && (
-        <FormItemEditor
-          selectedItem={selectedItem}
-          definition={formDefinition}
-          handleClose={() => setSelectedItem(undefined)}
-        />
-      )}
-      <Box
-        sx={
-          // Padding matches the padding usually applied in DashboardContentContainer.
-          // (It's moved in here because of the left drawer, above)
-          {
-            flexGrow: 1,
-            pt: 2,
-            pb: 8,
-            px: { xs: 1, sm: 3, lg: 4 },
-          }
-        }
-      >
-        <FormBuilderHeader
-          formDefinition={formDefinition}
-          lastUpdatedDate={lastUpdatedDate}
-        />
-        <Box sx={{ p: 4 }}>
-          <FormTree
-            definition={formDefinition.definition}
-            setSelectedItem={(item: FormItem) => setSelectedItem(item)}
-          />
-        </Box>
-        <Paper sx={{ p: 4 }}>
-          <Stack
-            direction='row'
-            justifyContent='space-between'
-            sx={{ alignItems: 'center' }}
-          >
-            <Stack direction='row' gap={2}>
-              <Button variant='outlined'>Save Draft</Button>
-              <Button>Publish</Button>
-            </Stack>
-            <Typography variant='body2'>
-              Last saved on {lastUpdatedDate} by {lastUpdatedBy}
-            </Typography>
-          </Stack>
-        </Paper>
-      </Box>
-    </Box>
+    <FormBuilder
+      formDefinition={formDefinition}
+      workingDefinition={workingDefinition}
+      setWorkingDefinition={setWorkingDefinition}
+      errorState={errorState}
+      lastUpdatedDate={lastUpdatedDate || undefined}
+      lastUpdatedBy={lastUpdatedBy}
+      selectedItem={selectedItem}
+      setSelectedItem={setSelectedItem}
+      originalLinkId={originalLinkId}
+      setOriginalLinkId={setOriginalLinkId}
```

Comment:
I think this passing around of "original link id" is a little confusing. If we only allow editing link id for new items -- before they are saved for the first time-- then this wouldn't be needed, correct? I think that would be preferable. Then we also don't have to worry about link id changes causing dependency problems. What do you think about adding that as a constraint? 

It doesn't look like we've yet implemented the behavior whereby when you add an item, the panel opens right away.  That would probably be necessary in order to implement the suggested constraint. (And the panel would have to "know" that it's working on a new item that has not yet been persisted).

Reply:
Digging through the code some more, another note is that we don't need to be passing the original link id all the way through the tree. The `FormEditorItemProperties` should already know what the "original" link ID is, because it is on the `item` that it receives initially. It could simply store that value:
`const [initialLinkId] = useState<string>(item.linkId);`

Reply:
Yes, this is way nicer. Thank you.

I created a new ticket https://github.com/open-path/Green-River/issues/6167 to capture the work for enforcing that constraint. It does seem nice because it lets us avoid some dependency problems. On the other hand, though, it might be frustrating to users (maybe us!) not to be able to update the link IDs of existing form items, especially items that only exist in a draft...

Reply:
Awesome thanks :) yeah I could see that, too.